### PR TITLE
[Form] deprecate integers as form names

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | 
| License       | MIT
| Doc PR        |

### CODE REVIEW

**Comment by xabbuh** [comment]
The `FormConfigBuilder` is the only class accepting integers as form names. All calling code in the component already passes strings. So this will affect few (if any) developers, but allows us to add a proper type hint in Symfony 5.

File: src/Symfony/Component/Form/FormConfigBuilder.php [hunk 1]
```diff
@@ -179,6 +179,10 @@ public function __construct($name, ?string $dataClass, EventDispatcherInterface
     {
         self::validateName($name);
 
+        if (\is_int($name)) {
```
Comment (linaori) [inline src/Symfony/Component/Form/FormConfigBuilder.php]: Is there a specific reason why strings with numerical values are still allowed?

File: UPGRADE-4.3.md [hunk 2]
```diff
@@ -23,6 +23,7 @@ Config
 Form
 ----
 
+ * Support for using integers as form names is deprecated and will lead to a type error in 5.0, use strings instead.
```
Comment (nicolas-grekas) [inline UPGRADE-4.3.md]: will it? automatic casting rules would turn the int to a string and all will be silent, no?

**Comment by nicolas-grekas** [comment]
What's the technical benefit of this change?

**Comment by nicolas-grekas** [comment]
I mean that I don't understand what this will change  allow from a technical pov: the name is explicitly cast to string a few lines below.
If we'll need to add more code to trigger the exception, I don't get the point :)
What did I miss?

**Comment by stof** [comment]
There won't be any TypeError in Symfony 5. There will be a cast (as done today explicitly in the constructor) because its only caller is FormBuilder (for the parent call), which is not running using strict types (as Symfony does not have them).

And even if we had typehints on that name everywhere in Symfony (where we already document `string` as the expected type), you would not get a TypeError unless your own code calling Symfony APIs is using strict_types (and then, we have no way to know that to decide whether we should trigger a deprecation or no)

**Comment by xabbuh** [comment]
Let me clarify my intention here: This change would not have any effect for our internal code as all the callers of this constructor inside the Form component already pass a string. This would only affect consumers of the component who build `FormConfigBuilder` instances on their own (and they would only get a type error if they use strict types).

But reiterating over my reasoning I think it's probably not worth the effort here. Let's close it.

**Pull request closed without merging**